packages/freezed/test/multiple_constructors_test.dart

```diff
@@ -504,8 +504,65 @@ void main() {
 }
 '''), throwsCompileError);
     });
+  });
+
+  group('SharedParamCommonSuperType', () {
+    test('has the common properties available', () {
+      SharedParamCommonSuperType value =
+          SharedParamCommonSuperType('a', 1337, 42);
+      expect(value.a, 'a');
+      expect(value.b, 1337);
+
+      value = SharedParamCommonSuperType.named('b', 2.4, 24);
+      expect(value.a, 'b');
+      expect(value.b, 2.4);
+    });
+
+    test('copy has different nullability available', () {
+      SharedParamCommonSuperType value =
+          SharedParamCommonSuperType('a', 1337, 42);
+      expect(value.a, 'a');
 
-    // TODO: shared property name but different type
+      value = value.copyWith(a: 'b');
+      expect(value.a, 'b');
+    });
+
+    test(
+      'copy has shared params with different nullability available as non-null',
+      () async {
+        await expectLater(compile(r'''
+import 'multiple_constructors.dart';
+
+void main() {
+  final param = SharedParamCommonSuperType('a', 1337, 42);
+  param.copyWith(a: '2');
+}
+'''), completes);
+
+        await expectLater(compile(r'''
+import 'multiple_constructors.dart';
+
+void main() {
+  final param = SharedParamCommonSuperType('a', 1337, 42);
+  param.copyWith(a: null);
+}
+'''), throwsCompileError);
+      },
+    );
+
+    test(
+      'copy does not have shared params with common super type available',
+      () async {
+        await expectLater(compile(r'''
+import 'multiple_constructors.dart';
+
+void main() {
+  final param = SharedParamCommonSuperType('a', 'b', 42);
+  param.copyWith(b: 42);
+}
```

Comment:
This probably could be done, (not a common super type, but by finding a common sub type).

For example in the null-case:
String?
String
common subtype == String, which is why copyWith a String makes sense because it works for all variations of the parameter.

For a not much more complicated hierarchy.
num
double
common subtype == double -> you can use double for the parameter

Finally, in general:
if (param is subtype of otherConstructorParam):
  commonSubtype = param
else if (otherConstructorParam is subtype of param):
  commonSubtype = otherConstructorParam
else
  commonSubtype = Never (no common subtype)


Additionally this test fails to compile not because of the copyWith but because of the String passed into the constructor I think

Reply:
The test was a bad rewrite of the earlier test for the different nullability. Yes it fails due to the incorrect string constructor paramter but it would also fail for the parameter which it was intended for.

I'll fix the test and also try to find a way of generally finding common subtypes.